lua/neogit/config.lua

```diff
@@ -1,157 +1,457 @@
-      ["f"] = "FetchPopup",
-      ["X"] = "ResetPopup",
-      ["M"] = "RemotePopup",
-      ["{"] = "GoToPreviousHunkHeader",
-      ["}"] = "GoToNextHunkHeader",
+    popup = {
+      kind = "split",
+    },
+    signs = {
+      hunk = { "", "" },
+      item = { ">", "v" },
+      section = { ">", "v" },
     },
-  },
-}
+    integrations = setmetatable({}, {
+      __index = function(_, key)
+        local ok, value = pcall(require, key)
+        return ok and value or false
+      end,
+    }),
+    sections = {
+      untracked = {
+        folded = false,
+      },
+      unstaged = {
+        folded = false,
+      },
+      staged = {
+        folded = false,
+      },
+      stashes = {
+        folded = true,
+      },
+      unpulled = {
+        folded = true,
+      },
+      unmerged = {
+        folded = false,
+      },
+      recent = {
+        folded = true,
+      },
+      rebase = {
+        folded = true,
+      },
+    },
+    ignored_settings = {
+      "NeogitPushPopup--force-with-lease",
+      "NeogitPushPopup--force",
+      "NeogitCommitPopup--allow-empty",
+      "NeogitRevertPopup--no-edit", -- TODO: Fix incompatible switches with default enables
+    },
+    mappings = {
+      finder = {
+        ["Select"] = { "<cr>" },
+        ["Close"] = { "<c-c>", "<esc>" },
+        ["Next"] = { "<c-n>", "<down>" },
+        ["Previous"] = { "<c-p>", "<up>" },
+        ["MultiselectToggleNext"] = { "<tab>" },
+        ["MultiselectTogglePrevious"] = { "<s-tab>" },
+        ["NOP"] = { "<c-j>" },
+      },
+      status = {
+        ["Close"] = "q",
+        ["InitRepo"] = "I",
+        ["Depth1"] = "1",
+        ["Depth2"] = "2",
+        ["Depth3"] = "3",
+        ["Depth4"] = "4",
+        ["Toggle"] = "<tab>",
+        ["Discard"] = "x",
+        ["Stage"] = "s",
+        ["StageUnstaged"] = "S",
+        ["StageAll"] = "<c-s>",
+        ["Unstage"] = "u",
+        ["UnstageStaged"] = "U",
+        ["DiffAtFile"] = "d",
+        ["CommandHistory"] = "$",
+        ["Console"] = "#",
+        ["RefreshBuffer"] = "<c-r>",
+        ["GoToFile"] = "<enter>",
+        ["VSplitOpen"] = "<c-v>",
+        ["SplitOpen"] = "<c-x>",
+        ["TabOpen"] = "<c-t>",
+        ["HelpPopup"] = "?",
+        ["DiffPopup"] = "D",
+        ["PullPopup"] = "p",
+        ["RebasePopup"] = "r",
+        ["MergePopup"] = "m",
+        ["PushPopup"] = "P",
+        ["CommitPopup"] = "c",
+        ["LogPopup"] = "L",
+        ["RevertPopup"] = "_",
+        ["StashPopup"] = "Z",
+        ["CherryPickPopup"] = "A",
+        ["BranchPopup"] = "b",
+        ["FetchPopup"] = "f",
+        ["ResetPopup"] = "X",
+        ["RemotePopup"] = "M",
+        ["GoToPreviousHunkHeader"] = "{",
+        ["GoToNextHunkHeader"] = "}",
+      },
+    },
+  }
+end
+
+M.values = M.get_default_values()
+
+---Validates the config
+---@return boolean false if the config is invalid
+---@return string all error messages emitted during validation
```

Comment:
You previously mentioned that it was hard to parse and programmatically use the output of the validate function in tests, especially if there are multiple errors.

We could strike *three* birds with one stone by making this return a *list* of errors. The current method, by returning a boolean for success, and a pretty human readable concatenated message for all the errors leads to some ambiguities. Given the function signature, it is entirely possible to return both `true` and errors at the same time.  Looking at the function from the outside. there would be no way to tell it could not do that, as it is entirely possible given the definition space, in which the output would be very malformed.

By returning a single list of strings, the success status of the validation becomes entirely self-evident by an empty list and it becomes impossible to end up in the trinary "on *and* off" state.

The header message could be added at the call site in setup as it is not really part of the validation. This way, the validation function becomes data oriented withing a simple well-defined space. There would be no ambiguities, and parsing and checking the validation message in the tests would be straight forward, one list element per error. This way, you could also validate that the *correct* error, and the right amount of them are emitted if you want to be pedantic about the tests.

This style of *constraining using definition* something that I've learned from use of Rust and Haskell and has worked wonders for large interconnected systems, just sharing my advice.

Reply:
No you're right, when I'm in scripting languages I have a habit of kludging things together. "Proper" programming languages tend to lend themselves to making this type of mistake more difficult. Rust especially makes it obvious on domain whereas Lua, at least for me, not so much.

Will be fixed here shortly. The only thing I'm thinking of doing is returning a key value store of them with the key containing the config value that is bad, say `signs.hunk` with the error message being its value. Might make it a bit quicker to test and validate.

Reply:
Oooh :eyes: 

Agreed, Rust is phenomenal at this.

```rust 
fn validate(config: Config) -> BTreeMap<ConfigKey, String>
```
 as a Rustacean would see it, love it!

Lua being lua though we'd need to remember to use `vim.tbl_count` as `#value` only considers integer keys

Reply:
> as #value only considers integer keys

I didn't know that... 👀. I have a bunch of stuff then that's uhhh, having a not so good time lmao.

Reply:
Oh no, seems like I poked the sleeping beast.

```vim
lua assert(#{a = 5}, 0)
```

This passes, lua truly is strange, and the list handling is *odd*

Reply:
That's cursed. Proof that we should be using elisp instead of lua. Anyhow, currently working on getting a `key -> { error_messages }` return going.

Reply:
Done @ten3roberts. Take a look and let me know on this specific change. I am still going to make modifications relevant to https://github.com/NeogitOrg/neogit/pull/629#discussion_r1271098487.

Of particular interest is how I'm doing the formatting now within the neogit setup function. You may want me to change that be a separate function that takes in those config errors and then outputs the string for error presentation. Just let me know.